app/components/submission/ManuscriptMetadata/SubjectAreaDropdown.test.js

```diff
@@ -0,0 +1,120 @@
+    })
+
+    const typeSubjectArea = letters => {
+      // We need 2 steps to programmatically simulate typing in a letter
+      // 1. Add a letter to the value of the input field (does not cause the dropdown to appear by itself)
+      selectWrapper.props().onInputChange(letters)
+      // 2. Open the dropdown
+      selectInput.simulate('change')
+    }
+
+    it('focusing causes the dropdown to open & display every option', () => {
+      typeSubjectArea('') // typing nothing into the field seems to be the only way to programmatically focus react-select)
+      expect(wrapper.find('Menu').exists()).toEqual(true)
+      expect(wrapper.find('Option')).toHaveLength(17)
+    })
+
+    it('an option can be selected from the dropdown', () => {
+      typeSubjectArea('b')
+      selectInput.simulate('keyDown', { keyCode: 9, key: 'Tab' })
+      expect(onChange.mock.calls).toHaveLength(1)
+      expect(wrapper.find('Menu').exists()).toEqual(false)
+      expect(wrapper.find('MultiValue')).toHaveLength(1) // MultiValue = a "tag" (i.e. selected option) in react-select
+      expect(
+        wrapper
+          .find('MultiValue')
+          .contains('Biochemistry and Chemical Biology'),
+      ).toBe(true)
+    })
+
+    it('2 options can be selected from the dropdown', () => {
+      typeSubjectArea('b')
+      selectInput.simulate('keyDown', { keyCode: 9, key: 'Tab' })
+      typeSubjectArea('c')
+      selectInput.simulate('keyDown', { keyCode: 9, key: 'Tab' })
+      expect(wrapper.find('MultiValue')).toHaveLength(2)
+      expect(wrapper.find('MultiValue').contains('Cancer Biology')).toBe(true)
+
+      // selectWrapper.props().onInputChange('E')
+      // selectInput.simulate('change')
+      // expect(wrapper.contains('Menu')).toBe(false)
+    })
+
+    it('selecting 2 options disables further selection', () => {
+      typeSubjectArea('b')
+      selectInput.simulate('keyDown', { keyCode: 9, key: 'Tab' })
+      typeSubjectArea('c')
+      selectInput.simulate('keyDown', { keyCode: 9, key: 'Tab' })
+      expect(wrapper.contains('DropdownIndicator')).toBe(false)
+    })
+
+    it('selection/deselection of 2 items triggers success message to show/hide', () => {
+      typeSubjectArea('b')
+      selectInput.simulate('keyDown', { keyCode: 9, key: 'Tab' })
+      typeSubjectArea('c')
+      selectInput.simulate('keyDown', { keyCode: 9, key: 'Tab' })
+      expect(wrapper.contains('No more than two subject areas')).toBe(true)
+
+      const crossOnFirstTag = wrapper.find('MultiValueRemove').at(0)
+      crossOnFirstTag.simulate('click')
+      expect(wrapper.find('MultiValue')).toHaveLength(1)
+      expect(wrapper.contains('No more than two subject areas')).toBe(false)
+    })
+
+    it('typing filters list of options based on first letter(s)', () => {
+      typeSubjectArea('c')
+      const totalStartingWithC = wrapper.find('Option').length
+      expect(totalStartingWithC).toBeLessThan(17)
+      wrapper.find('Option').forEach(option => {
```

Comment:
Why do you do this twice?
```
const wordsStartingWithC = wrapper.find('Option')
expect(wordsStartingWithC.length).toBeLessThan(17);
wordsStartingWithC.forEach(...)
```
Also, do we really need to assert that the length is less than the total number of options? It's really hardcoded, just checking that all the words in this list start with `c` or that this list is the same as the list of words starting with `c` from `eLifeOptions` (I guess building this second one here would work) seems enough.

Reply:
I need to call `wrapper.find('Option')` twice, because the number of `Option`s on the page changes after I simulate a new letter being typed. So this duplication is necessary:
```
const totalStartingWithC = wrapper.find('Option').length
// simulate click event
const totalStartingWithCe = wrapper.find('Option').length
```

If I wanted the variable name to stay the same I could have done this:
```
let totalOptions

totalOptions = wrapper.find('Option').length
// simulate click event
totalOptions = wrapper.find('Option').length
```

but I specifically wanted the 2 different lengths to be stored separately, so I could use them both in this assertion:
```
expect(totalStartingWithCe).toBeLessThan(totalStartingWithC)
```

But you're totally right. The `toBeLessThan` test is actually totally redundant, since I'm checking whether they start with the appropriate letters anyway. Thanks :+1: 


As for the rest, the assertions are not identical, so this isn't really duplication:
```
// 1st
expect(totalStartingWithC).toBeLessThan(17)
// 2nd
expect(totalStartingWithCe).toBeLessThan(totalStartingWithC)
```

and the `wrapper.find('Option').forEach(option => .....` is definitely duplication & could have been abstracted into a function. Seemed a bit overkill, but probably not & probably would have been far more readable :+1:

Reply:
That's not what I meant, sorry. I meant that you call `wrapper.find()` twice in a row for no reason (it's a function). It strictly refers to lines 102, 103 and 104. Look at what I was suggesting to use: call `wrapper.find()` once and store the result in one variable that you can use for both checks you do for letter `c`.

I was also suggesting that taking the `eLifeOptions` list that you defined in this file, filter it to have only words that start with `c` (writing the code for this) in this test, and then comparing with the output of `SubjectAreaDropdown` might be a good test.

Similar for `ce`.

Reply:
Yeah that's exactly what I thought too! But it doesn't seem to work like that.

If you save it in a variable, e.g. `const totalStartingWithC = wrapper.find()`, then simulate a change....
`totalStartingWithC` will still be equal to the _old_ value (i.e. 4)

You need to actually called `wrapper.find()` again.

Go figure :woman_shrugging:

Reply:
I don't see what you don't understand from what I was saying. Is this line simulating a change and I don't know about it?
`expect(totalStartingWithC).toBeLessThan(17)`